cmd/jujud/util/util.go

```diff
@@ -140,8 +138,47 @@ func ConnectionIsFatal(logger loggo.Logger, conns ...Pinger) func(err error) boo
 	}
 }
 
-// ConnectionIsDead returns true if the given pinger fails to ping.
-var ConnectionIsDead = func(logger loggo.Logger, conn Pinger) bool {
+// ConnectionIsDead returns true if the given Breakable is broken.
+var ConnectionIsDead = func(logger loggo.Logger, conn Breakable) bool {
+	select {
+	case <-conn.Broken():
```

Comment:
Ok, fair enough. How about I add IsBroken as suggested and have it check the Broken channel first, and if that succeeds it then tries a ping to be sure and avoid the race?

I disagree that pingConn is a hack. Ping is just an API call like any other. Why should it get a special method on api.Connection? That said, given that this broken your code, there's a risk that it'll break other places too. I might make it exported again given where we are with the 2.0 release.